exporter/exporterhelper/queued_retry_test.go

```diff
@@ -582,6 +582,60 @@ func TestQueuedRetryPersistenceEnabledStorageError(t *testing.T) {
 	require.Error(t, be.Start(context.Background(), host), "could not get storage client")
 }
 
+func TestQueuedRetryPersistentEnabled_shutdown_dataIsRequeued(t *testing.T) {
+	produceCounter := atomic.NewUint32(0)
+
+	req := newMockRequest(context.Background(), 3, errors.New("some error"))
+	be := dataRequeuedTest(t, req, produceCounter)
+
+	// shuts down and ensure the item is produced in the queue again
+	require.NoError(t, be.Shutdown(context.Background()))
+	assert.Eventually(t, func() bool {
+		return produceCounter.Load() == uint32(2)
+	}, time.Second, 1*time.Millisecond)
+}
+
+func dataRequeuedTest(t *testing.T, req internal.Request, produceCounter *atomic.Uint32) *baseExporter {
+	qCfg := NewDefaultQueueSettings()
+	qCfg.NumConsumers = 1
+	rCfg := NewDefaultRetrySettings()
+	rCfg.InitialInterval = time.Millisecond
+	rCfg.MaxElapsedTime = 0 // retry infinitely so shutdown can be triggered
+
+	be, err := newBaseExporter(defaultSettings, fromOptions(WithRetry(rCfg), WithQueue(qCfg)), "", nopRequestUnmarshaler())
+	require.NoError(t, err)
+
+	// we start correctly with a file storage extension
+	require.NoError(t, be.Start(context.Background(), &mockHost{}))
+
+	// wraps original queue so we can count operations
+	be.qrSender.queue = &producerConsumerQueueWithCounter{
+		ProducerConsumerQueue: be.qrSender.queue,
+		produceCounter:        produceCounter,
+	}
+	be.qrSender.requeuingEnabled = true
+
+	// replace nextSender inside retrySender to always return error so it doesn't exit send loop
+	castedSender, ok := be.qrSender.consumerSender.(*retrySender)
+	require.True(t, ok, "consumerSender should be a retrySender type")
+	castedSender.nextSender = &ErrorRequestSender{
+		errToReturn: errors.New("some error"),
+	}
+
+	sendErrChan := make(chan error, 1)
+	go func() {
+		sendErrChan <- be.sender.send(req)
+	}()
+	require.NoError(t, <-sendErrChan)
```

Comment:
Why channel?

```suggestion
	require.NoError(t, be.sender.send(req))
```

Reply:
The channel can block until the send operation completes in the goroutine. We can also extract the error returned and verify the error is nil.

Reply:
`require.NoError(t, be.sender.send(req))` is a blocking call and also verifies the error

Reply:
@MovieStoreGuy suggested in [thread](https://github.com/open-telemetry/opentelemetry-collector/pull/6815#discussion_r1068754947) that if the program exits before the goroutine completes, the error can cause a panic in the test program. Having the channel blocking outside can help.

As for the need to send the data in a goroutine is that the send operation puts the handling in an for loop. So we have to call `send` in `shutdown` in different goroutine.

Reply:
But why was the goroutine needed in the first place?

Reply:
The `send` operation has a for loop to keep processing the request until the processing has permanently failed or the program is exiting. While our test wants to test out the shutdown path, which is triggered when `shutdown` is called.

So if we don't put `send` in a different goroutine, the test program can't call `shutdown`. As the call to `send` will be waiting until it finishes but it won't finish until `shutdown` is called.

Reply:
> As the call to send will be waiting until it finishes but it won't finish until shutdown is called.

But it finishes right after spawning the goroutine, you're waiting for it to be finished with no error `require.NoError(t, <-sendErrChan)`. This asynchronous block doesn't make sense. It does the same work as the one-liner that I suggested

Reply:
Please make sure your test does what you expect

Reply:
After double checking the code, it seems like i have confused `queuedRetrySender.Send` with `retrySender.Send`. In the test when `be.sender.send(req)` is called, it invokes `queuedRetrySender.Send`, which puts the request on a channel so the consumer will poll and use `retrySender` to `Send`.

I will update the call to not use goroutine. Apology for the confusion.